lib/tesla/middleware/retry.ex

```diff
@@ -84,15 +92,26 @@ defmodule Tesla.Middleware.Retry do
   defp retry(env, next, context) do
     res = Tesla.run(env, next)
 
-    if context.should_retry.(res) do
-      backoff(context.max_delay, context.delay, context.retries, context.jitter_factor)
-      context = update_in(context, [:retries], &(&1 + 1))
-      retry(env, next, context)
-    else
-      res
+    {:arity, should_retry_arity} = :erlang.fun_info(context.should_retry, :arity)
+
+    cond do
+      should_retry_arity == 1 and context.should_retry.(res) ->
+        do_retry(env, next, context)
+
+      should_retry_arity == 2 and context.should_retry.(res, env) ->
```

Comment:
```suggestion
      should_retry_arity == 2 and context.should_retry.(res, env, context) ->
```

We should also pass the `context,` so the function may be 3 arities now reading the code.